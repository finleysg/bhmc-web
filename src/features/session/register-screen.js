import "./session.scss"

import React from "react"

import { useAuth } from "context/auth-context"
import { RegisterForm } from "features/session/register-form"
import { MdAccountCircle } from "react-icons/md"

function RegisterScreen() {
  const { register } = useAuth()

  return (
    <div className="login">
      <div className="login__block active">
        <div className="login__header bg-blue">
          <i>
            <MdAccountCircle />
          </i>
          Create an Account
        </div>

        <div className="login__body">
          <RegisterForm onSubmit={register} />
        </div>
      </div>
    </div>
  )
}

export { RegisterScreen }
