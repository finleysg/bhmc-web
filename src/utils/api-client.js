import { queryCache } from "react-query"

const serverURL = process.env.REACT_APP_SERVER_URL

async function client(endpoint, { data, token, headers: customHeaders, ...customConfig } = {}) {
  const config = {
    method: data ? "POST" : "GET",
    body: data ? JSON.stringify(data) : undefined,
    headers: {
      Authorization: token ? `Token ${token}` : undefined,
      "Content-Type": data ? "application/json" : undefined,
      ...customHeaders,
    },
    ...customConfig,
  }

  const parseError = (error) => {
    if (typeof error === "string") {
      return error
    }
    if (error.non_field_errors) {
      return error.non_field_errors[0]
    }
    // TODO
    return error
  }

  return window.fetch(`${serverURL}/${endpoint}`, config).then(async (response) => {
    if (response.status === 401) {
      queryCache.clear()
      // await auth.logout()
      // refresh the page for them
      window.location.assign(window.location)
      return Promise.reject({ message: "Please re-authenticate." })
    }
    if (response.ok) {
      if (response.status !== 204) {
        const data = await response.json()
        return data
      }
      return null
    } else {
      const data = await response.json()
      return Promise.reject(parseError(data))
    }
  })
}

export { client }
