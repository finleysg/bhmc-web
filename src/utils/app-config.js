export const currentEnvironment = +process.env.REACT_APP_CURRENT_ENVIRONMENT
export const stripePublicKey = process.env.REACT_APP_STRIPE_PUBLIC_KEY
export const giphyApiKey = process.env.REACT_APP_GIPHY_API_KEY
export const seniorRateAge = +process.env.REACT_APP_SENIOR_AGE
export const seniorCompetitionAge = +process.env.REACT_APP_SENIOR_COMPETITION_AGE
export const sentryApiKey = process.env.REACT_APP_SENTRY_API_KEY
export const version = process.env.REACT_APP_VERSION

export const maintenanceMode = process.env.REACT_APP_MODE !== "Live"
