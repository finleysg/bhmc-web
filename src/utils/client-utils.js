const API_BASE = process.env.REACT_APP_API_URL
const AUTH_BASE = process.env.REACT_APP_AUTH_URL
const SERVER_BASE = process.env.REACT_APP_SERVER_URL

const normalizeEndpoint = (endpoint) => {
  const [base, querystring] = endpoint.split("?")
  if (!base.startsWith("/") && base.endsWith("/")) {
    return `${base}${querystring ? "?" + querystring : ""}`
  }
  if (!base.startsWith("/") && !base.endsWith("/")) {
    return `${base}/${querystring ? "?" + querystring : ""}`
  }
  if (base.startsWith("/") && base.endsWith("/")) {
    return `${base.substring(1)}${querystring ? "?" + querystring : ""}`
  }
  if (base.startsWith("/") && !base.endsWith("/")) {
    return `${base.substring(1)}/${querystring ? "?" + querystring : ""}`
  }
  throw new Error("Someone messed up their url munging!")
}

const apiUrl = (endpoint) => {
  return `${API_BASE}/${normalizeEndpoint(endpoint)}`
}
const authUrl = (endpoint) => {
  return `${AUTH_BASE}/${normalizeEndpoint(endpoint)}`
}
const serverUrl = (endpoint) => {
  return `${SERVER_BASE}/${normalizeEndpoint(endpoint)}`
}

const parseError = (error) => {
  if (typeof error === "string") {
    return error
  }
  if (error.non_field_errors) {
    return error.non_field_errors[0]
  }
  if (error.detail) {
    return error.detail
  }
  return JSON.stringify(error)
}

export { apiUrl, authUrl, parseError, serverUrl }
