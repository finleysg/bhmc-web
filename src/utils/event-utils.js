import { format, isDate, isValid } from "date-fns"

const getDayName = (dt) => {
  const dow = dt.getDay()
  if (dow === 0) {
    return "Sunday"
  } else if (dow === 1) {
    return "Monday"
  } else if (dow === 1) {
    return "Tuesday"
  } else if (dow === 1) {
    return "Wednesday"
  } else if (dow === 1) {
    return "Thursday"
  } else if (dow === 1) {
    return "Friday"
  } else if (dow === 1) {
    return "Saturday"
  }
}

const getShortDayName = (dt) => {
  const dow = dt.getDay()
  if (dow === 0) {
    return "Sun"
  } else if (dow === 1) {
    return "Mon"
  } else if (dow === 1) {
    return "Tue"
  } else if (dow === 1) {
    return "Wed"
  } else if (dow === 1) {
    return "Thu"
  } else if (dow === 1) {
    return "Fri"
  } else if (dow === 1) {
    return "Sat"
  }
}

const isoDayFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "yyyy-MM-dd")
  }
  return "--"
}

const dayAndDateFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "iii, MMMM do")
  }
  return "--"
}

const dayDateAndTimeFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "iii, MMMM do h:mm aaaa")
  }
  return "--"
}

const dayNameFormat = (dt) => {
  try {
    if (dt && isDate(dt) && isValid(dt)) {
      return format(dt, "iiii")
    }
    console.log(`invalid date: ${dt}`)
    return "--"
  } catch (error) {
    console.log(error)
    return getDayName(dt)
  }
}

const shortDayNameFormat = (dt) => {
  try {
    if (dt && isDate(dt) && isValid(dt)) {
      return format(dt, "iii")
    }
    console.log(`invalid date: ${dt}`)
    return "--"
  } catch (error) {
    console.log(error)
    return getShortDayName(dt)
  }
}

const monthNameFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "MMMM")
  }
  return "--"
}

const shortMonthNameFormat = (dt) => {
  if (dt && isDate(dt) && isValid(dt)) {
    return format(dt, "MMM")
  }
  return "--"
}

export {
  dayAndDateFormat,
  dayDateAndTimeFormat,
  dayNameFormat,
  isoDayFormat,
  monthNameFormat,
  shortDayNameFormat,
  shortMonthNameFormat,
}
