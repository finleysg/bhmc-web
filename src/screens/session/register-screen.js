import "./session.scss"

import React from "react"

import { useAuth } from "context/auth-context"
import { MdAccountCircle } from "react-icons/md"
import { RegisterForm } from "screens/session/register-form"

function RegisterScreen() {
  const { register } = useAuth()

  return (
    <div className="login">
      <div className="login__block active">
        <div className="login__header bg-blue">
          <i>
            <MdAccountCircle />
          </i>
          Create an Account
        </div>

        <div className="login__body">
          <RegisterForm onSubmit={register} />
        </div>
      </div>
    </div>
  )
}

export { RegisterScreen }
