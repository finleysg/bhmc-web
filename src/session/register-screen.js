import "./session.scss"

import React from "react"

import { useAuth } from "context/auth-context"
import { MdAccountCircle } from "react-icons/md"
import { RegisterForm } from "session/register-form"
import { RoutingMenu } from "session/routing-menu"

function RegisterScreen() {
  const { register } = useAuth()

  return (
    <div className="login">
      <div className="login__block active">
        <div className="login__header bg-blue">
          <i>
            <MdAccountCircle />
          </i>
          Create an Account
          <RoutingMenu
            links={[
              { to: "/session/login", name: "Login" },
              { to: "/session/reset-password", name: "Reset My Password" },
              { to: "/home", name: "Home" },
            ]}
          />
        </div>

        <div className="login__body">
          <RegisterForm onSubmit={register} />
        </div>
      </div>
    </div>
  )
}

export { RegisterScreen }
