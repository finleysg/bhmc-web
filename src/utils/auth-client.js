import { authUrl, parseError } from "./client-utils"

async function client(endpoint, data, token) {
  const config = {
    method: data ? "POST" : "GET",
    body: data ? JSON.stringify(data) : undefined,
    headers: {
      Authorization: token ? `Token ${token}` : undefined,
      "Content-Type": data ? "application/json" : undefined,
    },
  }

  const url = authUrl(endpoint)

  return window.fetch(url, config).then(async (response) => {
    if (response.ok) {
      if (response.status !== 204) {
        const data = await response.json()
        return data
      }
      return null
    } else {
      const data = await response.json()
      return Promise.reject(parseError(data))
    }
  })
}

export { client }
