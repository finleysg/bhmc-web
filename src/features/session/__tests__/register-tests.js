import { act, render, screen, waitFor } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import React from "react"

import { rest } from "msw"
import { setupServer } from "msw/node"
import { handlers } from "test/auth-handlers"
import { buildRegisterForm } from "test/generate/auth"
import { AuthWrapper, deferred } from "test/test-utils"
import { authUrl } from "utils/client-utils"

import { RegisterScreen } from "../register-screen"

const server = setupServer(...handlers)

const mockNav = jest.fn()
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNav,
}))

beforeAll(() => {
  server.listen()
})
afterAll(() => server.close())
afterEach(() => {
  server.resetHandlers()
  jest.clearAllMocks()
})

test("successful registration", async () => {
  const { promise, resolve } = deferred()

  render(
    <AuthWrapper>
      <RegisterScreen />
    </AuthWrapper>,
  )

  expect(screen.getByLabelText(/loading/i)).toBeInTheDocument()

  await act(() => {
    resolve()
    return promise
  })

  const { first_name, last_name, email, password, re_password } = buildRegisterForm()

  userEvent.type(screen.getByRole("textbox", { name: /first name/i }), first_name)
  userEvent.type(screen.getByRole("textbox", { name: /last name/i }), last_name)
  userEvent.type(screen.getByRole("textbox", { name: /email/i }), email)
  userEvent.type(screen.getByLabelText("Password", { exact: true }), password)
  userEvent.type(screen.getByLabelText(/confirm password/i), re_password)
  userEvent.click(screen.getByRole("button"))

  await waitFor(() => expect(mockNav).toHaveBeenCalledWith("session/account/confirm"))
})

test("duplicate email displays custom message", async () => {
  const registerUrl = authUrl("users")
  server.use(
    rest.post(registerUrl, async (req, res, ctx) => {
      const err = ctx.json(["user already exists"])
      return res(ctx.status(400), err)
    }),
  )

  const { promise, resolve } = deferred()

  render(
    <AuthWrapper>
      <RegisterScreen />
    </AuthWrapper>,
  )

  expect(screen.getByLabelText(/loading/i)).toBeInTheDocument()

  await act(() => {
    resolve()
    return promise
  })

  const { first_name, last_name, email, password, re_password } = buildRegisterForm()

  userEvent.type(screen.getByRole("textbox", { name: /first name/i }), first_name)
  userEvent.type(screen.getByRole("textbox", { name: /last name/i }), last_name)
  userEvent.type(screen.getByRole("textbox", { name: /email/i }), email)
  userEvent.type(screen.getByLabelText("Password", { exact: true }), password)
  userEvent.type(screen.getByLabelText(/confirm password/i), re_password)
  userEvent.click(screen.getByRole("button"))

  const alert = await screen.findByRole("alert")

  expect(alert.textContent).toMatchSnapshot()
  expect(screen.getByRole("link", { name: /reset your password/i })).toBeInTheDocument()
})
