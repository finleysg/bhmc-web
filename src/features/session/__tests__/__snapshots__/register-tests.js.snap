// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`duplicate email displays custom message 1`] = `"\\"We already have an account with that email. Do you need to reset your password?\\""`;
